plans/init.pp

```diff
@@ -0,0 +1,98 @@
+# @param reboot_delay How long (in seconds) to wait before rebooting. Defaults to 0, reboot immediately.
+#        Note on most Linux platforms, this time can only be specified in minutes, so non-0 values will
+#        delay reboot by at least 1 minute.
+# @param disconnect_wait How long (in seconds) to wait before checking whether the server has rebooted. Defaults to 1.
+# @param reconnect_wait How long (in seconds) to attempt to reconnect before giving up. Defaults to 120.
+# @param retry_interval How long (in seconds) to wait between retries. Defaults to 1.
+plan reboot (
+  TargetSpec $nodes,
+  Optional[String] $message = undef,
+  Integer[0] $reboot_delay = 0,
+  Integer[0] $disconnect_wait = 1,
+  Integer[0] $reconnect_wait = 120,
+  Integer[0] $retry_interval = 1,
+) {
+  $start_time = Timestamp()
+  $targets = get_targets($nodes)
+
+  # Get last boot time
+  $begin_boot_time_results = without_default_logging() || {
+    run_task('reboot::last_boot_time', $targets)
+  }
+
+  # Reboot; catch errors here because the connection may get cut out from underneath
+  $expect_disconnect = ($reboot_delay == 0)
+  $reboot_result = run_task('reboot', $nodes, timeout => $reboot_delay, message => $message, _catch_errors => $expect_disconnect)
+
+  # Wait long enough for all targets to trigger reboot, plus disconnect_wait to allow for shutdown time.
+  # Errored results should only occur if timeout was 0, so we assume they rebooted immediately.
+  $timeouts = $reboot_result.ok_set.map |$result| { $result['timeout'] }
+  $wait = max($timeouts + [0])
+  reboot::sleep($wait+$disconnect_wait)
+
+  # Wait for reboot in a loop
+  ## Check if we can connect; if we can retrieve last boot time.
+  ## Mark finished for targets with a new last boot time.
+  ## If we still have targets check for timeout, sleep if not done.
+  $failed = without_default_logging() || {
+    $reconnect_wait.reduce($targets) |$down, $_| {
+      if $down.empty() {
+        break()
+      }
+
+      $current_boot_time_results = run_task('reboot::last_boot_time', $down, _catch_errors => true)
+
+      # Compare boot times
+      $failed_results = $current_boot_time_results.filter |$current_boot_time_res| {
+        # If this one errored, need to check it again
+        if !$current_boot_time_res.ok() {
+          true
+          } else {
+            # If this succeeded, then we have a boot time, compare it against the begin_boot_time
+            $target_name = $current_boot_time_res.target().name()
+            $begin_boot_time_res = $begin_boot_time_results.find($target_name)
+
+            # If the boot times are the same, then we need to check it again
+            $current_boot_time_res.value() == $begin_boot_time_res.value()
+          }
+      }
+
+      # $failed_results is an array of results, turn it into a ResultSet so we can
+      # extract the targets from it
+      $failed_targets = ResultSet($failed_results).targets()
```

Comment:
If someone was to not set a `disconnect_wait`, but their servers actually take 2min to disconnect. This would immediately return success and the plan would proceed, only for the servers to die later. It really need to explicitly wait and ensure that each server goes from connected -> disconnected -> connected

Reply:
I think you misunderstand what the plan is doing. It's using 'last boot time' to determine when the server has restarted. This is a more robust method of checking than waiting for the server to disconnect then reconnect.

However, I am considering what that looks like in the console, where we'd have lots of failed runs of the `last_boot_time` task.

Reply:
I think the plan behavior - several failed runs of a task - is an ok trade-off for robustness.

Reply:
Yes I do misunderstand, sorry. However I do think the UX in the console is going to be nasty. Unless bolt could pass to the console when it had handled an error, so it was like a less-red failure, or a "handled failure" or something.

Reply:
Must have been a red herring in my testing then